### ISSUE TITLE
Assert error in h2_do_req(), http2/cache_http2_proto.c line 493

### ISSUE THREAD

**Issue by simonvik** [varnishcache/varnish-cache#2305]
```
Panic at: Sat, 08 Apr 2017 16:06:41 GMT
Assert error in h2_do_req(), http2/cache_http2_proto.c line 493:
  Condition(req->req_body_status == REQ_BODY_NONE) not true.
version = varnish-5.1.1 revision de38712, vrt api = 6.0
ident = Linux,4.4.30-20161130-1423-0a2a05d,x86_64,-junix,-smalloc,-smalloc,-hcritbit,epoll
now = 1648807.363910 (mono), 1491667599.027457 (real)
Backtrace:
  0x438c25: /usr/sbin/varnishd() [0x438c25]
  0x460839: /usr/sbin/varnishd() [0x460839]
  0x4520a2: /usr/sbin/varnishd() [0x4520a2]
  0x45254b: /usr/sbin/varnishd() [0x45254b]
  0x7fc79bf006ba: /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7fc79bf006ba]
  0x7fc79bc3682d: /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7fc79bc3682d]
thread = (cache-worker)
thr.req = 0x7fc204c06020 {
  vxid = 105614567, transport = H2
  step = R_STP_LOOKUP,
  req_body = R_BODY_WITH_LEN,
  restarts = 0, esi_level = 0,
  sp = 0x7fc4dfc1ba20 {
    fd = 323, vxid = 105614565,
    t_open = 1491667598.302321,
    t_idle = 1491667598.347864,
    transport = H2 {
      streams {
        0x00000000 idle  
        0x00000001 c-rem 
        0x00000003 c-rem 
        0x00000005 c-rem 
        0x00000007 c-rem 
      }
    }
    client = to: ********* 53192,
    privs = 0x7fc4dfc1ba88 {
    },
  },
  ws = 0x7fc204c06208 {
    id = \"req\",
    {s, f, r, e} = {0x7fc204c08008, +1616, +122864, +122864},
  },
  http_conn = 0x7fc204c06130 {
    doclose = NULL,
    ws = (nil) {
    },
    {rxbuf_b, rxbuf_e} = {(nil), (nil)},
    {pipeline_b, pipeline_e} = {(nil), (nil)},
    content_length = 0,
    body_status = none,
    first_byte_timeout = 0.000000,
    between_bytes_timeout = 0.000000,
  },
  http[req] = 0x7fc204c062a0 {
    ws = 0x7fc204c06208 {
      [Already dumped, see above]
    },
    hdrs {
      \"POST\",
      \"/series/\",
      \"HTTP/2.0\",
      \"host: to: *********\",
      \"scheme: https\",
      \"content-length: 4\",
      \"origin: https://www.google.pt\",
      \"user-agent: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36\",
      \"ping-to: *********\",
      \"content-type: text/ping\",
      \"accept: */*\",
      \"accept-language: pt-PT,pt;q=0.8,en-US;q=0.6,en;q=0.4\",
      \"X-Forwarded-For: *********\",
      \"*********: *********\",
      \"********* https\",
      \"X-Forwarded-Proto: https\",
      \"to: *********: to: *********\",
      \"X-UA-Device: pc\",
      \"*********: pc\",
      \"Accept-Encoding: gzip\",
    },
  },
  vcl = {
    name = \"vcl_045ce3ae-5156-42ef-a0e0-2e6b4f6c196a\",
    busy = 3113,
    discard = 0,
    state = auto,
    temp = warm,
    conf = {
      srcname = {
          *********
      },
    },
  },
  vmods = {
    std = {Varnish 5.1.1 de38712, 6.0},
    one = {Varnish 5.1.1 de38712, 6.0},
    maxminddb = {Varnish 5.1.1 de38712, 6.0},
    directors = {Varnish 5.1.1 de38712, 6.0},
    xkey = {Varnish 5.1.1 de38712, 6.0},
    header = {Varnish 5.1.1 de38712, 6.0},
    cookie = {Varnish 5.1.1 de38712, 6.0},
    vsthrottle = {Varnish 5.1.1 de38712, 6.0},
    tcp = {Varnish 5.1.1 de38712, 6.0},
    bodyaccess = {Varnish 5.1.1 de38712, 6.0},
  },
  flags = {
  },
},
thr.busyobj = (nil) {
},
```

Varnish 5.1.1 patched to what ended up being 5.1.2
Running on ubuntu 16.04

Attached nm-output from varnish-binary
[nm.txt](https://github.com/varnishcache/varnish-cache/files/907701/nm.txt)

**Comment by bsdphk** [comment]
0x438c25: /usr/sbin/varnishd() [0x438c25] - pan_ic()
  0x460839: /usr/sbin/varnishd() [0x460839] - h2_do_req() (near the end)
  0x4520a2: /usr/sbin/varnishd() [0x4520a2] - WRK_Thread()

**Comment by simonvik** [comment]
Unsure if this is 100% related but: 

`nghttp -H ":method:POST" -H 'content-length:4' http://localhost -v`
Makes the request hang a long time before it gives up too.

Other servers seems to return GOAWAY as soon as they receive content-length and 
headers with END_STREAM set as they dont expect to get DATA.

Google returns PROTOCOL_ERROR and nginx sends a readable error message : "400 Bad Request"

**Comment by nigoroll** [comment]
I accidentally referenced this issue in a commit related to #2405

**Comment by daghf** [comment]
Hi @simonvik 

I think the original crash reported was fixed in 4460caa89.

As for your second point on `END_STREAM` + `Content-Length` I'm struggling to find an RFC reference for how we should fail.

The RFC does say 

```
   END_STREAM (0x1):  When set, bit 0 indicates that the header block
      (Section 4.3) is the last that the endpoint will send for the
      identified stream.
```

which is obviously conflicting with a `Content-Length` header being present. Coming up with a fix would be simple enough, but I'm not sure if the proper fix would be a `GOAWAY` for that stream in particular or a `PROTOCOL_ERROR` for the connection as a whole.

@bsdphk @Dridi or others, ideas ?

**Comment by daghf** [comment]
I think I might have spoken too soon regarding this being fixed via 4460caa. I'm able to trigger the same panic when we reembark from waitinglist with a request body.

I'll cook up a fix
